lib/Install.js

```diff
@@ -1,86 +1,71 @@
 // @flow
 
-const child_process = require('child_process');
+const spawn = require('cross-spawn');
 const fs = require('fs');
 const path = require('path');
 const rimraf = require('rimraf');
-const Compiler = require('./Compile');
+const ElmJson = require('./ElmJson');
+const Project = require('./Project');
 
-function install(pathToElmBinary /*: string */, packageName /*: string */) {
-  var oldSourceDirectories;
+void Project;
 
-  const dirPath = path.join(
-    Compiler.getGeneratedCodeDir(process.cwd()),
-    'install'
-  );
+function install(
+  project /*: typeof Project.Project */,
+  pathToElmBinary /*: string */,
+  packageName /*: string */
+) /*: 'SuccessfullyInstalled' | 'AlreadyInstalled' */ {
+  const generatedCodeDir = path.join(project.generatedCodeDir, 'install');
 
   try {
     // Recreate the directory to remove any artifacts from the last time
     // someone ran `elm-test install`. We do not delete this directory after
     // the installation finishes in case the user needs to debug the test run.
-    if (fs.existsSync(dirPath)) {
+    if (fs.existsSync(generatedCodeDir)) {
       // We can replace this with `fs.rmdirSync(dir, { recursive: true })`
       // once Node.js 10 is EOL 2021-04-30 and support for Node.js 10 is dropped.
-      rimraf.sync(dirPath);
+      rimraf.sync(generatedCodeDir);
     }
-    fs.mkdirSync(dirPath, { recursive: true });
-  } catch (err) {
-    console.error(
-      'Unable to create temporary directory for elm-test install.',
-      err
+    fs.mkdirSync(generatedCodeDir, { recursive: true });
+  } catch (error) {
+    throw new Error(
+      `Unable to create temporary directory for elm-test install: ${error.message}`
     );
-    process.exit(1);
   }
 
-  var elmJson = JSON.parse(fs.readFileSync('elm.json', 'utf8'));
-  var tmpElmJsonPath = path.join(dirPath, 'elm.json');
-  var isPackage;
-
-  switch (elmJson['type']) {
-    case 'package':
-      isPackage = true;
-      break;
+  const { elmJson } = project;
 
-    case 'application':
-      isPackage = false;
-      break;
-
-    default:
-      console.error('Unrecognized elm.json type:', elmJson['type']);
-      process.exit(1);
-  }
-
-  // This mirrors the behavior of `elm install` passing a package that is
-  // already installed. Say it's already installed, then exit 0.
   if (
-    (isPackage && elmJson['test-dependencies'].hasOwnProperty(packageName)) ||
-    (!isPackage &&
-      elmJson['test-dependencies']['direct'].hasOwnProperty(packageName))
+    elmJson.type === 'package'
+      ? elmJson['test-dependencies'].hasOwnProperty(packageName)
+      : elmJson['test-dependencies'].direct.hasOwnProperty(packageName)
   ) {
-    console.log('It is already installed!');
-    return;
+    return 'AlreadyInstalled';
```

Comment:
Returning instead of logging here is nice because when in this scenario:

1. `elm-test init`
2. Delete tests/
3. `elm-test init` again, to regenerate tests/

Now, “It is already installed!” won’t be logged in that case. That message was only ever meant for `elm-test install user/package` when `user/package` was already installed.

Reply:
Because `init` calls `install` to add the test package as a dev dep. Yep :+1: